File: src/main/java/aurocosh/divinefavor/common/item/talisman_container/spell_bow/capability/SpellBowProvider.java
package aurocosh.divinefavor.common.item.talisman_container.spell_bow.capability;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.ICapabilitySerializable;
import net.minecraftforge.items.CapabilityItemHandler;

public class SpellBowProvider implements ICapabilitySerializable<NBTTagCompound> {
    ISpellBowHandler instance = SpellBowDataHandler.CAPABILITY_SPELL_BOW.getDefaultInstance();

    @Override
    public boolean hasCapability(Capability<?> capability, EnumFacing facing) {
        if(capability == SpellBowDataHandler.CAPABILITY_SPELL_BOW)
            return true;
        return capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY;
    }

    @Override
    public <T> T getCapability(Capability<T> capability, EnumFacing facing) {
        if(!hasCapability(capability,facing))
            return null;
        if(capability == SpellBowDataHandler.CAPABILITY_SPELL_BOW)
            return SpellBowDataHandler.CAPABILITY_SPELL_BOW.cast(instance);
        return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast(instance.getStackHandler());
    }

    @Override
    public NBTTagCompound serializeNBT() {
        return (NBTTagCompound) SpellBowDataHandler.CAPABILITY_SPELL_BOW.getStorage().writeNBT(SpellBowDataHandler.CAPABILITY_SPELL_BOW, instance, null);
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        SpellBowDataHandler.CAPABILITY_SPELL_BOW.getStorage().readNBT(SpellBowDataHandler.CAPABILITY_SPELL_BOW, instance, null, nbt);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/talisman_tools/spell_bow/capability/SpellBowProvider.kt
package aurocosh.divinefavor.common.item.talisman_tools.spell_bow.capability

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilitySerializable
import net.minecraftforge.items.CapabilityItemHandler

class SpellBowProvider : ICapabilitySerializable<NBTTagCompound> {
    internal var instance = SpellBowDataHandler.CAPABILITY_SPELL_BOW.defaultInstance

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        return if (capability === SpellBowDataHandler.CAPABILITY_SPELL_BOW) true else capability === CapabilityItemHandler.ITEM_HANDLER_CAPABILITY
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        if (!hasCapability(capability, facing))
            return null
        return if (capability === SpellBowDataHandler.CAPABILITY_SPELL_BOW) SpellBowDataHandler.CAPABILITY_SPELL_BOW.cast<T>(instance) else CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast<T>(instance!!.getStackHandler())
    }

    override fun serializeNBT(): NBTTagCompound {
        val storage = SpellBowDataHandler.CAPABILITY_SPELL_BOW.storage
        return storage.writeNBT(SpellBowDataHandler.CAPABILITY_SPELL_BOW, instance, null) as NBTTagCompound
    }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        val storage = SpellBowDataHandler.CAPABILITY_SPELL_BOW.storage
        storage.readNBT(SpellBowDataHandler.CAPABILITY_SPELL_BOW, instance, null, nbt)
    }
}
